File: idea-plugin/src/main/kotlin/com/gitai/commit/CommitMessageFormatter.kt
package com.gitai.commit

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper

object CommitMessageFormatter {
    private const val maxLength = 80
    private val objectMapper = ObjectMapper()
    private val subjectPattern = Regex("""(?i)^[a-z][a-z0-9-]*(?:\([^)]+\))?:\s+\S.*$""")
    private val conventionalPrefixPattern = Regex("""(?i)^([a-z][a-z0-9-]*(?:\([^)]+\))?):\s*(\S.*)$""")

    fun format(message: String): String {
        val unwrapped = unwrapModelResponse(message)
        if (unwrapped.isBlank()) return ""

        val subject = unwrapped.lineSequence()
            .map { cleanLine(it) }
            .filter { it.isNotBlank() }
            .firstOrNull { looksLikeSubject(it) || looksLikeShortChineseSentence(it) }

        val normalized = subject
            ?.let { normalizeSubject(it, unwrapped) }
            ?: summarizeVerbose(unwrapped)
            ?: normalizeSubject(cleanLine(unwrapped.lineSequence().firstOrNull { it.isNotBlank() }.orEmpty()), unwrapped)

        return ensureConventionalPrefix(normalized).take(maxLength).trimEnd()
    }

    private fun unwrapModelResponse(message: String): String {
        val withoutFences = message.lineSequence()
            .map { it.trim() }
            .filter { !it.startsWith("```") }
            .joinToString("\n")
            .trim()

        if (!withoutFences.startsWith("{")) return withoutFences

        return runCatching {
            val json = objectMapper.readTree(withoutFences)
            extractJsonText(json).orEmpty()
        }.getOrElse { withoutFences }
    }

    private fun extractJsonText(json: JsonNode): String? =
        listOf("commit_message", "message", "response")
            .firstNotNullOfOrNull { key ->
                json.path(key).asText("").trim().takeIf { isUsableModelText(it) }
            }

    private fun isUsableModelText(text: String): Boolean {
        val lower = text.lowercase()
        return text.isNotBlank() && lower !in setOf("success", "ok", "done", "true", "false", "null", "none")
    }

    private fun cleanLine(line: String): String =
        line.removePrefix("# ")
            .removePrefix("> ")
            .removePrefix("- ")
            .removePrefix("* ")
            .replace(Regex("""^\d+[.)]\s*"""), "")
            .replace(Regex("""\s+"""), " ")
            .trim()

    private fun looksLikeSubject(line: String): Boolean = subjectPattern.matches(line)

    private fun looksLikeShortChineseSentence(line: String): Boolean =
        line.any(::isChineseCharacter) &&
            line.length <= maxLength &&
            !line.endsWith(":") &&
            !line.endsWith("：")

    private fun summarizeVerbose(text: String): String? {
        val lower = text.lowercase()
        val sentenceCount = Regex("""[.!?。！？]+""").findAll(text).count()
        val numberedLineCount = text.lineSequence().count { it.trim().matches(Regex("""\d+[.)]\s+.*""")) }
        val hasVerboseMarker = listOf(
            "this commit",
            "the diff shows",
            "the changes include",
            "here's a breakdown",
            "overall,",
            "i have reviewed",
            "the repository now contains"
        ).any { lower.contains(it) }

        if (!hasVerboseMarker && !(text.length > 120 && sentenceCount >= 2) && numberedLineCount < 2) {
            return null
        }

        return summarizeByKeywords(lower)
    }

    private fun normalizeSubject(subject: String, fullText: String): String {
        val prefix = conventionalPrefixPattern.find(subject)?.groupValues?.get(1)
        val stripped = stripConventionalPrefix(subject)
        if (stripped.any(::isChineseCharacter)) {
            val body = if (isGenericPluginSubject(stripped)) {
                defaultPluginFeatureSummary()
            } else {
                stripped
            }
            return ensureConventionalPrefix(body, prefix)
        }

        val translated = translateEnglishSubject(stripped)
        if (translated.any(::isChineseCharacter)) return ensureConventionalPrefix(translated, prefix)

        return ensureConventionalPrefix(summarizeByKeywords(fullText.lowercase()), prefix)
    }

    private fun translateEnglishSubject(subject: String): String {
        translateSpecificEnglishSubject(subject)?.let { return it }

        var translated = subject
            .replace("generated commit messages", "生成的提交信息", ignoreCase = true)
            .replace("generated commit message", "生成的提交信息", ignoreCase = true)
            .replace("commit messages", "提交信息", ignoreCase = true)
            .replace("commit message", "提交信息", ignoreCase = true)
            .replace("commit", "提交信息", ignoreCase = true)
            .replace("output language", "输出语言", ignoreCase = true)
            .replace("openai-compatible", "OpenAI 兼容", ignoreCase = true)
            .replace("provider", "提供商", ignoreCase = true)
            .replace("settings", "设置", ignoreCase = true)
            .replace("switching", "切换", ignoreCase = true)
            .replace("switch", "切换", ignoreCase = true)
            .replace("support", "支持", ignoreCase = true)
            .replace("add", "添加", ignoreCase = true)
            .replace("update", "更新", ignoreCase = true)
            .replace("improve", "优化", ignoreCase = true)
            .replace("fix", "修复", ignoreCase = true)
            .replace("refactor", "重构", ignoreCase = true)
            .replace("formatting", "格式化", ignoreCase = true)
            .replace("analysis", "分析", ignoreCase = true)
            .replace("analyzer", "分析器", ignoreCase = true)
            .replace("filtering", "过滤", ignoreCase = true)
            .replace("filter", "过滤器", ignoreCase = true)
            .replace("tests", "测试", ignoreCase = true)
            .replace("test", "测试", ignoreCase = true)
            .replace("english", "英文", ignoreCase = true)
            .replace("chinese", "中文", ignoreCase = true)
            .replace(Regex("""\s+"""), "")
            .trim()

        translated = translated.removeSuffix(".")
        return translated
    }

    private fun translateSpecificEnglishSubject(subject: String): String? {
        val normalized = subject.lowercase()

        return when {
            normalized.contains("add log files for indexing diagnostics and open-telemetry metrics") ||
                normalized.contains("add log files for indexing diagnostics and open telemetry metrics") ->
                "添加索引诊断日志和OpenTelemetry指标文件"

            else -> null
        }
    }

    private fun summarizeByKeywords(lower: String): String = when {
        lower.contains("english") && lower.contains("chinese") && lower.contains("commit message") ->
            "支持中英文提交信息切换"

        featureSummary(lower) != null ->
            featureSummary(lower)!!

        lower.contains("model provider") || lower.contains("openai-compatible") ->
            "完善模型提供商支持"

        lower.contains("diff") ->
            "完善差异分析能力"

        else ->
            "优化提交信息生成"
    }

    private fun isGenericPluginSubject(subject: String): Boolean =
        subject.contains("Git AI Commit", ignoreCase = true) &&
            listOf("核心功能", "主要功能", "基础功能").any { subject.contains(it) }

    private fun defaultPluginFeatureSummary(): String =
        "新增提交信息格式化、差异分析、Git diff 过滤、模型提供商管理和测试覆盖"

    private fun featureSummary(lower: String): String? {
        val features = mutableListOf<String>()

        if (listOf("commitmessageformatter", "commit message formatting", "formatting commit messages")
                .any { lower.contains(it) }
        ) {
            features += "提交信息格式化"
        }
        if (lower.contains("json") || lower.contains("response")) {
            features += "JSON 解析"
        }
        if (listOf("diff analysis", "diff analyzer", "analyzing diffs").any { lower.contains(it) }) {
            features += "差异分析"
        }
        if (listOf("git diff filtering", "gitdifffilter", "diff filter").any { lower.contains(it) }) {
            features += "Git diff 过滤"
        }
        if (listOf("model provider", "provider management", "provider selection").any { lower.contains(it) }) {
            features += "模型提供商管理"
        }
        if (lower.contains("openai-compatible")) {
            features += "OpenAI 兼容客户端"
        }
        if (listOf("promptbuilder", "prompt style", "prompt template").any { lower.contains(it) }) {
            features += "提示词优化"
        }
        if (listOf("test coverage", "unit tests", "html reports", "report pages").any { lower.contains(it) }) {
            features += "测试覆盖"
        }

        if (features.size < 2) return null

        return "新增" + when (features.size) {
            1 -> features[0]
            2 -> features.joinToString("和")
            else -> features.dropLast(1).joinToString("、") + "和" + features.last()
        }
    }

    private fun stripConventionalPrefix(subject: String): String =
        subject.replace(Regex("""(?i)^[a-z][a-z0-9-]*(?:\([^)]+\))?:\s*"""), "").trim()

    private fun ensureConventionalPrefix(message: String, preferredPrefix: String? = null): String {
        val trimmed = message.trim()
        if (trimmed.isBlank()) return ""
        if (conventionalPrefixPattern.matches(trimmed)) return trimmed

        val prefix = preferredPrefix?.takeIf { it.isNotBlank() } ?: "feat"
        return "$prefix: $trimmed"
    }

    private fun isChineseCharacter(char: Char): Boolean = char in '\u4e00'..'\u9fff'
}

File: vscode-plugin/src/core/commitMessageFormatter.ts
/**
 * Commit message formatter matching the Kotlin `CommitMessageFormatter`.
 * Post-processes the LLM response: unwraps JSON, strips markdown,
 * translates English to Chinese, ensures Conventional Commit prefix.
 */
export class CommitMessageFormatter {
    private static readonly MAX_LENGTH = 80;
    private static readonly SUBJECT_PATTERN = /^[a-z][a-z0-9-]*(?:\([^)]+\))?:\s+\S.*$/i;
    private static readonly CONVENTIONAL_PREFIX_PATTERN = /^([a-z][a-z0-9-]*(?:\([^)]+\))?):\s*(\S.*)$/i;

    format(message: string, style: 'short' | 'detailed' = 'short'): string {
        const unwrapped = this.unwrapModelResponse(message);
        if (!unwrapped) {
            return '';
        }

        if (style === 'detailed') {
            return this.formatDetailed(unwrapped);
        }

        return this.formatShort(unwrapped);
    }

    private formatShort(unwrapped: string): string {
        const lines = unwrapped.split('\n');
        const subject = lines
            .map(l => this.cleanLine(l))
            .filter(l => l.length > 0)
            .find(l => this.looksLikeSubject(l) || this.looksLikeShortChineseSentence(l));

        let normalized: string;
        if (subject) {
            normalized = this.normalizeSubject(subject, unwrapped);
        } else {
            const summary = this.summarizeVerbose(unwrapped);
            if (summary) {
                normalized = summary;
            } else {
                const firstLine = lines.find(l => l.trim().length > 0) ?? '';
                normalized = this.normalizeSubject(this.cleanLine(firstLine), unwrapped);
            }
        }

        return this.ensureConventionalPrefix(normalized).slice(0, CommitMessageFormatter.MAX_LENGTH).trimEnd();
    }

    private formatDetailed(unwrapped: string): string {
        const lines = unwrapped.split('\n');
        const subjectIndex = this.findSubjectIndex(lines, unwrapped);
        const subjectSource = subjectIndex >= 0
            ? this.cleanLine(lines[subjectIndex])
            : this.cleanLine(lines.find(l => l.trim().length > 0) ?? '');
        const normalizedSubject = this.normalizeSubject(subjectSource, unwrapped);
        const bodyLines = this.buildDetailedBody(lines, subjectIndex, unwrapped);

        if (bodyLines.length === 0) {
            return this.formatShort(unwrapped);
        }

        return [this.ensureConventionalPrefix(normalizedSubject).slice(0, CommitMessageFormatter.MAX_LENGTH).trimEnd(), '', ...bodyLines].join('\n').trimEnd();
    }

    private unwrapModelResponse(message: string): string {
        const withoutFences = message.split('\n')
            .map(l => l.trim())
            .filter(l => !l.startsWith('```'))
            .join('\n')
            .trim();

        if (!withoutFences.startsWith('{')) {
            return withoutFences;
        }

        try {
            const json = JSON.parse(withoutFences);
            const text = this.extractJsonText(json);
            return text ?? '';
        } catch {
            return withoutFences;
        }
    }

    private extractJsonText(json: Record<string, unknown>): string | null {
        for (const key of ['commit_message', 'message', 'response']) {
            const value = json[key];
            if (typeof value === 'string' && this.isUsableModelText(value)) {
                return value;
            }
        }
        return null;
    }

    private isUsableModelText(text: string): boolean {
        const lower = text.toLowerCase();
        return text.length > 0 && !['success', 'ok', 'done', 'true', 'false', 'null', 'none'].includes(lower);
    }

    private cleanLine(line: string): string {
        return line
            .replace(/^# /, '')
            .replace(/^> /, '')
            .replace(/^- /, '')
            .replace(/^\* /, '')
            .replace(/^\d+[.)]\s*/, '')
            .replace(/\s+/g, ' ')
            .trim();
    }

    private looksLikeSubject(line: string): boolean {
        return CommitMessageFormatter.SUBJECT_PATTERN.test(line);
    }

    private looksLikeShortChineseSentence(line: string): boolean {
        return this.hasChineseCharacter(line) &&
            line.length <= CommitMessageFormatter.MAX_LENGTH &&
            !line.endsWith(':') &&
            !line.endsWith('：');
    }

    private findSubjectIndex(lines: string[], fullText: string): number {
        const subjectPatterns = lines
            .map(l => this.cleanLine(l))
            .map((line, index) => ({ line, index }))
            .find(({ line }) => line.length > 0 && (this.looksLikeSubject(line) || this.looksLikeShortChineseSentence(line)));

        if (subjectPatterns) {
            return subjectPatterns.index;
        }

        const firstNonBlank = lines.findIndex(l => l.trim().length > 0);
        return firstNonBlank >= 0 ? firstNonBlank : (this.summarizeVerbose(fullText) ? 0 : -1);
    }

    private buildDetailedBody(lines: string[], subjectIndex: number, fullText: string): string[] {
        const bodyLines = lines
            .slice(Math.max(subjectIndex + 1, 0))
            .map(line => this.cleanLine(line))
            .filter(line => line.length > 0)
            .map(line => this.toBulletLine(line))
            .filter((line): line is string => line.length > 0);

        if (bodyLines.length > 0) {
            return bodyLines;
        }

        const summary = this.summarizeVerbose(fullText) ?? this.summarizeByKeywords(fullText.toLowerCase());
        if (!summary) {
            return [];
        }

        return [`- ${summary}`];
    }

    private toBulletLine(line: string): string {
        const cleaned = this.stripBulletMarker(line);
        if (!cleaned) {
            return '';
        }

        return cleaned.startsWith('- ') ? cleaned : `- ${cleaned}`;
    }

    private stripBulletMarker(line: string): string {
        return line
            .replace(/^[-*•]\s*/, '')
            .replace(/^\d+[.)]\s*/, '')
            .replace(/\s+/g, ' ')
            .trim();
    }

    private summarizeVerbose(text: string): string | null {
        const lower = text.toLowerCase();
        const sentenceCount = (text.match(/[.!?。！？]+/g) ?? []).length;
        const numberedLineCount = text.split('\n').filter(l => /^\d+[.)]\s+/.test(l.trim())).length;
        const verboseMarkers = [
            'this commit',
            'the diff shows',
            'the changes include',
            "here's a breakdown",
            'overall,',
            'i have reviewed',
            'the repository now contains',
        ];
        const hasVerboseMarker = verboseMarkers.some(m => lower.includes(m));

        if (!hasVerboseMarker && !(text.length > 120 && sentenceCount >= 2) && numberedLineCount < 2) {
            return null;
        }

        return this.summarizeByKeywords(lower);
    }

    private normalizeSubject(subject: string, fullText: string): string {
        const prefixMatch = CommitMessageFormatter.CONVENTIONAL_PREFIX_PATTERN.exec(subject);
        const prefix = prefixMatch?.[1];
        const stripped = this.stripConventionalPrefix(subject);

        if (this.hasChineseCharacter(stripped)) {
            const body = this.isGenericPluginSubject(stripped)
                ? this.defaultPluginFeatureSummary()
                : stripped;
            return this.ensureConventionalPrefix(body, prefix);
        }

        const translated = this.translateEnglishSubject(stripped);
        if (this.hasChineseCharacter(translated)) {
            return this.ensureConventionalPrefix(translated, prefix);
        }

        return this.ensureConventionalPrefix(
            this.summarizeByKeywords(fullText.toLowerCase()),
            prefix,
        );
    }

    private translateEnglishSubject(subject: string): string {
        const specific = this.translateSpecificEnglishSubject(subject);
        if (specific) {
            return specific;
        }

        const replacements: Array<[RegExp, string]> = [
            [/generated commit messages/gi, '生成的提交信息'],
            [/generated commit message/gi, '生成的提交信息'],
            [/commit messages/gi, '提交信息'],
            [/commit message/gi, '提交信息'],
            [/commit/gi, '提交信息'],
            [/output language/gi, '输出语言'],
            [/openai-compatible/gi, 'OpenAI 兼容'],
            [/provider/gi, '提供商'],
            [/settings/gi, '设置'],
            [/switching/gi, '切换'],
            [/switch/gi, '切换'],
            [/support/gi, '支持'],
            [/add/gi, '添加'],
            [/update/gi, '更新'],
            [/improve/gi, '优化'],
            [/fix/gi, '修复'],
            [/refactor/gi, '重构'],
            [/formatting/gi, '格式化'],
            [/analysis/gi, '分析'],
            [/analyzer/gi, '分析器'],
            [/filtering/gi, '过滤'],
            [/filter/gi, '过滤器'],
            [/tests/gi, '测试'],
            [/test/gi, '测试'],
            [/english/gi, '英文'],
            [/chinese/gi, '中文'],
        ];

        let translated = subject;
        for (const [pattern, replacement] of replacements) {
            translated = translated.replace(pattern, replacement);
        }
        translated = translated.replace(/\s+/g, '');
        translated = translated.replace(/\.$/, '');
        return translated.trim();
    }

    private translateSpecificEnglishSubject(subject: string): string | null {
        const normalized = subject.toLowerCase();

        if (
            normalized.includes('add log files for indexing diagnostics and open-telemetry metrics') ||
            normalized.includes('add log files for indexing diagnostics and open telemetry metrics')
        ) {
            return '添加索引诊断日志和OpenTelemetry指标文件';
        }

        return null;
    }

    private summarizeByKeywords(lower: string): string {
        if (lower.includes('english') && lower.includes('chinese') && lower.includes('commit message')) {
            return '支持中英文提交信息切换';
        }

        const featureSummary = this.featureSummary(lower);
        if (featureSummary) {
            return featureSummary;
        }

        if (lower.includes('model provider') || lower.includes('openai-compatible')) {
            return '完善模型提供商支持';
        }

        if (lower.includes('diff')) {
            return '完善差异分析能力';
        }

        return '优化提交信息生成';
    }

    private isGenericPluginSubject(subject: string): boolean {
        return /Git AI Commit/i.test(subject) &&
            ['核心功能', '主要功能', '基础功能'].some(w => subject.includes(w));
    }

    private defaultPluginFeatureSummary(): string {
        return '新增提交信息格式化、差异分析、Git diff 过滤、模型提供商管理和测试覆盖';
    }

    private featureSummary(lower: string): string | null {
        const features: string[] = [];

        if (['commitmessageformatter', 'commit message formatting', 'formatting commit messages']
            .some(k => lower.includes(k))) {
            features.push('提交信息格式化');
        }
        if (lower.includes('json') || lower.includes('response')) {
            features.push('JSON 解析');
        }
        if (['diff analysis', 'diff analyzer', 'analyzing diffs'].some(k => lower.includes(k))) {
            features.push('差异分析');
        }
        if (['git diff filtering', 'gitdifffilter', 'diff filter'].some(k => lower.includes(k))) {
            features.push('Git diff 过滤');
        }
        if (['model provider', 'provider management', 'provider selection'].some(k => lower.includes(k))) {
            features.push('模型提供商管理');
        }
        if (lower.includes('openai-compatible')) {
            features.push('OpenAI 兼容客户端');
        }
        if (['promptbuilder', 'prompt style', 'prompt template'].some(k => lower.includes(k))) {
            features.push('提示词优化');
        }
        if (['test coverage', 'unit tests', 'html reports', 'report pages'].some(k => lower.includes(k))) {
            features.push('测试覆盖');
        }

        if (features.length < 2) {
            return null;
        }

        const prefix = '新增';
        if (features.length === 1) {
            return prefix + features[0];
        }
        if (features.length === 2) {
            return prefix + features.join('和');
        }
        return prefix + features.slice(0, -1).join('、') + '和' + features[features.length - 1];
    }

    private stripConventionalPrefix(subject: string): string {
        return subject.replace(/^[a-z][a-z0-9-]*(?:\([^)]+\))?:\s*/i, '').trim();
    }

    private ensureConventionalPrefix(message: string, preferredPrefix?: string): string {
        const trimmed = message.trim();
        if (!trimmed) {
            return '';
        }
        if (CommitMessageFormatter.CONVENTIONAL_PREFIX_PATTERN.test(trimmed)) {
            return trimmed;
        }

        const prefix = (preferredPrefix && preferredPrefix.length > 0) ? preferredPrefix : 'feat';
        return `${prefix}: ${trimmed}`;
    }

    private hasChineseCharacter(str: string): boolean {
        return /[\u4e00-\u9fff]/.test(str);
    }
}
